chirp/drivers/kgq10h.py

```diff
@@ -0,0 +1,3203 @@
+            if pol == "R":
+                val += 0x2000
+            return val
+
+        rx_mode = tx_mode = None
+        rxtone = txtone = 0x0000
+
+        if mem.tmode == "Tone":
+            tx_mode = "Tone"
+            rx_mode = None
+            txtone = int(mem.rtone * 10) + 0x8000
+        elif mem.tmode == "TSQL":
+            rx_mode = tx_mode = "Tone"
+            rxtone = txtone = int(mem.ctone * 10) + 0x8000
+        elif mem.tmode == "DTCS":
+            tx_mode = rx_mode = "DTCS"
+            txtone = _set_dcs(mem.dtcs, mem.dtcs_polarity[0])
+            rxtone = _set_dcs(mem.dtcs, mem.dtcs_polarity[1])
+        elif mem.tmode == "Cross":
+            tx_mode, rx_mode = mem.cross_mode.split("->")
+            if tx_mode == "DTCS":
+                txtone = _set_dcs(mem.dtcs, mem.dtcs_polarity[0])
+            elif tx_mode == "Tone":
+                txtone = int(mem.rtone * 10) + 0x8000
+            if rx_mode == "DTCS":
+                rxtone = _set_dcs(mem.rx_dtcs, mem.dtcs_polarity[1])
+            elif rx_mode == "Tone":
+                rxtone = int(mem.ctone * 10) + 0x8000
+
+        _mem.rxtone = rxtone
+        _mem.txtone = txtone
+
+        LOG.debug("Set TX %s (%i) RX %s (%i)" %
+                  (tx_mode, _mem.txtone, rx_mode, _mem.rxtone))
+
+    def set_memory(self, mem):
+        number = mem.number
+
+        _mem = self._memobj.memory[number]
+        _nam = self._memobj.names[number]
+
+        if mem.empty:
+            _mem.set_raw("\x00" * (_mem.size() // 8))
+            self._memobj.valid[number] = 0
+            self._memobj.names[number].set_raw("\x00" * (_nam.size() // 8))
+            return
+
+        _mem.rxfreq = int(mem.freq / 10)
+        if mem.duplex == "off":
+            # _mem.txfreq = 0xFFFFFFFF
+            _mem.txfreq = 0x00000000
+        elif mem.duplex == "split":
+            _mem.txfreq = int(mem.offset / 10)
+        elif mem.duplex == "off":
+            for i in range(0, 4):
+                _mem.txfreq[i].set_raw("\xFF")
+        elif mem.duplex == "+":
+            _mem.txfreq = int(mem.freq / 10) + int(mem.offset / 10)
+        elif mem.duplex == "-":
+            _mem.txfreq = int(mem.freq / 10) - int(mem.offset / 10)
+        else:
+            _mem.txfreq = int(mem.freq / 10)
+        _mem.scan_add = int(mem.skip != "S")
+
+        if ((mem.mode == "FM") or (mem.mode == "NFM")):
+            _mem.iswide = int((mem.mode == "FM"))
+            _mem.am_mode = 0
+        elif mem.mode == "AM":
+            _mem.am_mode = 1
+
+        # set the tone
+        self._set_tone(mem, _mem)
+
+        # set the power
+        _mem.power = _mem.power & 0x3
+        if mem.power:
+            if _mem.power > 3:
+                _mem.power = 3
+            _mem.power = self.POWER_LEVELS.index(mem.power)
+        else:
+            _mem.power = True
+
+        for setting in mem.extra:
+            setattr(_mem, setting.get_name(), setting.value)
```

Comment:
Right here, when we get to `am_mode`, you set it right back to 0 because the memory's `am_mode` hasn't yet been set. Since you set it to 0 here, then when the UI refreshes the memory after the set operation, your own `get_memory()` is returning `mode=FM`.

If you're doing the edits direct in the spreadsheet with "show extra", you can't set both fields to a consistent value in one operation, but you can in the properties dialog, which lets you make several changes that get sent to the driver all at once.

You can fix the driver to not be broken like this, but I think it's probably not ideal. I think you should probably not make am_mode be a tri-state thing including "Off" and instead just make it be `["RX", "RX+TX"]`. That way the user selects `mode=AM` and `am_mode` separately, which is how it should be anyway. Since `am_mode` can't be stored when FM is selected anyway, if they choose anything for `am_mode` when mode is not AM, that's an error or something to ignore.

You could also only populate am_mode with "off" if mode is set to FM/NFM and populate it with the two choices only when mode=AM. That way there's nothing to select for am_mode *until* mode=AM.

Make sense?

Reply:
One concern with this approach.   There is only a single setting in memory to determine AM/FM.   It is the am_mode setting.
it has 3 values, Off, AM Rx, AM Rx+Tx.   If am_mode is OFF, then FM mode is used.  So when you say don't make am_mode a tri-state, it is exactly a tri-state setting.   What I was trying to do is when FM is selected in the spreadsheet view -  force am_mode to OFF like the radio expects.    When AM is selected, I wanted to force am_mode to AM Rx.  Then if the user wanted AM Rx+Tx they would choose that in the properties box.   I don't know how to not use am_mode as a tri-state.   What am I missing?

Reply:
What you're describing is basically my primary suggestion above: if `mode==FM`, then am_mode is off regardless. No matter what they choose for am_mode in the properties or spreadsheet view, you ignore that if `mode==FM`. Then if `mode==AM`, you default to whichever am_mode makes sense, and then let them choose between those two sub-modes. It's the same behavior as was discussed on the list for the FT450, fwiw.

Reply:
I am confused.. The code does exactly what we are talking about.  But then after executing the IF / ELIF statement on MODE, and setting  am_mode correctly,

        if ((mem.mode == "FM") or (mem.mode == "NFM")):
            _mem.iswide = int((mem.mode == "FM"))
            _mem.am_mode = 0
        elif mem.mode == "AM":
            _mem.am_mode = 1

the following code is run 
        for setting in mem.extra:
              setattr(_mem, setting.get_name(), setting.value)
  
which then sets am_mode back to what it was previously.   

If I remove the code
        for setting in mem.extra:
              setattr(_mem, setting.get_name(), setting.value)

then the mem.mode value is updated as I want it to be through the spreadsheet UI, but then I can't change any of the Extra settings and get them to take effect.

What am I missing that gets the 2 pieces of code to play nice together?

Reply:
I added this to the code and it seems to be doing what I want.
When mode != AM -  the mem.extra am_mode is not shown in the properties and it is forced to 0.
When mode = AM, the mem.extra allows selection of AM Rx or AM Rx+Tx
Switching from FM/NFM to AM or AM to FM/NFM must be done via the spreadsheet UI.
The properties diaglog box can only change AM modes when AM is selected via the spreadsheet UI mem.mode setting.

        for setting in mem.extra:
            if setting.get_name() != "am_mode":
                setattr(_mem, setting.get_name(), setting.value)
            else:
                if mem.mode != "AM":
                    setattr(_mem, setting.get_name(), 0)
                elif int(setting.value) == 2:
                    setattr(_mem, setting.get_name(), 2)

Although I think I need to add another elif to handle switching from value 2 back to default value 1.

                elif int(setting.value) == 1:
                    setattr(_mem, setting.get_name(), 1)

Reply:
with the above changes - all now pass Tox which was previously failing form FM/NFM/AM mode switching.

  style: OK (64.62=setup[27.91]+cmd[36.72] seconds)
  unit: OK (41.86=setup[12.97]+cmd[28.89] seconds)
  driver: OK (557.69=setup[13.22]+cmd[544.47] seconds)
  congratulations :) (664.42 seconds)